airflow/jobs/backfill_job_runner.py

```diff
@@ -588,61 +588,81 @@ def _per_task_process(key, ti: TaskInstance, session):
-                            )
+                            if task.max_active_tis_per_dag is not None:
+                                num_running_task_instances_in_task = DAG.get_num_task_instances(
+                                    dag_id=self.dag_id,
+                                    task_ids=[task.task_id],
+                                    states=self.STATES_COUNT_AS_RUNNING,
+                                    session=session,
+                                )
 
-                            if num_running_task_instances_in_task_dagrun >= task.max_active_tis_per_dagrun:
-                                raise TaskConcurrencyLimitReached(
-                                    "Not scheduling since Task concurrency per DAG run limit is reached."
+                                if num_running_task_instances_in_task >= task.max_active_tis_per_dag:
+                                    raise TaskConcurrencyLimitReached(
+                                        "Not scheduling since Task concurrency limit is reached."
+                                    )
+
+                            if task.max_active_tis_per_dagrun is not None:
+                                num_running_task_instances_in_task_dagrun = DAG.get_num_task_instances(
+                                    dag_id=self.dag_id,
+                                    run_id=ti.run_id,
+                                    task_ids=[task.task_id],
+                                    states=self.STATES_COUNT_AS_RUNNING,
+                                    session=session,
                                 )
 
-                        _per_task_process(key, ti, session)
-                        session.commit()
+                                if (
+                                    num_running_task_instances_in_task_dagrun
+                                    >= task.max_active_tis_per_dagrun
+                                ):
+                                    raise TaskConcurrencyLimitReached(
+                                        "Not scheduling since Task concurrency per DAG run limit is reached."
+                                    )
+
+                            _per_task_process(key, ti, session)
+                            try:
+                                session.commit()
+                                # break the retry loop
+                                break
+                            except OperationalError:
+                                self.log.error(
+                                    "Failed to commit task state due to operational error. "
+                                    "The job will retry this operation so if your backfill succeeds, "
+                                    "you can safely ignore this message.",
+                                    exc_info=True,
+                                )
+                                session.rollback()
+                                # outer loop will retry
```

Comment:
is there a missing check for the current attempt number?
```suggestion
                                if i = max_attempts - 1:
                                    raise
                                # outer loop will retry
```

Reply:
> is there a missing check for the current attempt number?

I don't think so . Why ?

Reply:
Ah yeah

Reply:
I thought it's better to just skip the backfill - the error is a bit less scary then but yes, likely better to reraise it

Reply:
I was under the impression initially that when I skip it, backfill will attempt it again with outer loop but it was wrong.

Reply:
Fixed - also changed the comment to reflect that it is **just** loop, not the outer one as I thought that will take care of retry initially (that was really wrong assumption).